src/rust-2021/prelude.md

```diff
@@ -92,26 +194,56 @@ mod submodule {
 }
 
 // `MyTrait` isn't in scope here and can only be referred to through the path `submodule::MyTrait`
+// `MyTrait` はここではスコープ内になく、パス付きで `submodule::MyTrait` としか利用できない
 fn bar(f: Box<dyn submodule::MyTrait>) {
   // If `std::convert::TryInto` is in scope (as would be the case in Rust 2021),
   // then it becomes ambiguous which `try_into` method to call
+  // `std::convert::TryInto` がスコープ内にあるときは (Rust 2021 ではそうなのだが)、
+  // どちらの `try_into` メソッドを呼び出せばいいかが曖昧になる
   f.try_into();
 }
 ```
 
+<!--
 Unlike with static dispatch methods, calling a trait method on a trait object does not require that the trait be in scope. The code above works 
 as long as there is no trait in scope with a conflicting method name. When the `TryInto` trait is in scope (which is the case in Rust 2021),
 this causes an ambiguity. Should the call be to `MyTrait::try_into` or `std::convert::TryInto::try_into`?
+-->
+
+静的ディスパッチのときと違って、トレイトオブジェクトに対してトレイトメソッドを呼び出すときは、そのトレイトがスコープ内にある必要はありません。
+`TryInto` トレイトがスコープ内にあるときは (Rust 2021 ではそうなのですが)、曖昧性が発生します。
+`MyTrait::try_into` と `std::convert::TryInto::try_into` のどちらが呼び出されるべきなのでしょうか？
 
+<!--
 In these cases, we can fix this by adding an additional dereferences or otherwise clarify the type of the method receiver. This ensures that 
 the `dyn Trait` method is chosen, versus the methods from the prelude trait. For example, turning `f.try_into()` above into `(&*f).try_into()` 
 ensures that we're calling `try_into` on the `dyn MyTrait` which can only refer to the `MyTrait::try_into` method.
+-->
 
+この場合、さらなる参照外しをするか、もしくはメソッドレシーバーの型を明示することで修正できます。
+これにより、`dyn Trait` のメソッドとプレリュードのトレイトのメソッドのどちらが選ばれているかが明確になります。
+たとえば、上の `f.try_into()` を `(&*f).try_into()` にすれば、<!-- 意訳。ここの部分の意味がわかりません-->確実に `MyTrait::try_into` を指し示しているとわかるので<!--ここまで-->、
+`dyn MyTrait` の方の `try_into` が呼び出されていると決定されます。
```

Comment:
```suggestion
たとえば、上の `f.try_into()` を `(&*f).try_into()` にすると、`dyn MyTrait` に対して `try_into` を呼び出すことがはっきりします。
これに該当するのは`MyTrait::try_into`メソッドのみです。
```

たしかに以下のようなコードを書いたらエラーになりました。

```rust
pub mod submodule {
    pub trait MyTrait {
        fn try_into(&self);
    }
}

fn bar(f: Box<dyn submodule::MyTrait>) {
  f.try_into();
  // error[E0283]: type annotations needed
  //   = note: cannot satisfy `_: TryFrom<Box<dyn MyTrait>>`
  //   = note: required because of the requirements on the impl of `TryInto<_>` for `Box<dyn MyTrait>`
  // help: use the fully qualified path for the potential candidates
  //   |
  // 8 |   <T as TryFrom<U>>::try_into(f);
  //   |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  // 8 |   <u32 as TryFrom<i8>>::try_into(f);
  //   |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  // 8 |   <u32 as TryFrom<isize>>::try_into(f);
  //   |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  // 8 |   <u32 as TryFrom<i128>>::try_into(f);
  //   |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //     and 213 other candidates
}
```

そして`f.try_into();`を`(*f).try_into();`や`(&*f).try_into();`に変えたらエラーが解消しました。

`f`は`Box<dyn MyTrait>`型なので、`*`で参照外しをすると`dyn MyTrait`型になります。その状態で`try_into`を呼ぶと、コンパイラーは`MyTrait::try_into`を選択するようです。（`dyn トレイト`のトレイトメソッドは固有メソッドとして扱われる。 https://rustc-dev-guide.rust-lang.org/method-lookup.html#candidate-assembly）